### DISCUSSION TITLE
Cobbler 3.2.2 on ALmaLinux 8.5  - dropping out to grub prompt

### DISCUSSION THREAD

**Discussion by Fluffyrodent** [cobbler/cobbler#2876] [Version specific questions]
Hi,

I've been struggling for weeks to get a cobbler server to install both a legacy centos7 machine and an EFI centos7 (secure and non-secure) machine on baremetal and HyperV - so 4 scenario types effectively.

I think I am misunderstanding something about the mkgrub for the targets.

I did it for legacy successfully on a centos7 cobbler (2.8.5) server using pxelinux.0 for both a physical legacy machine and a HyperV legacy machine but EFI failed to load so switched to an Alma 8.5 server to get a more up to date version of cobbler (3.2.2) and it mostly works but there is a small final config I'm not getting I think.

Secure boot is off for EFI at this stage of testing in my set up so no shim/shimx64 involved yet.

Here is my installation process.  I apologise for the size of this post, I wanted to get it all in place as it's taken me ages to work it out and it just maybe might help someone else as procedures are scattered and erratic online I found.

Main questions are at the bottom in the grub section.

Thank you very much.

-Stephen

# New Cobbler Server Installation

ALMA...8.5 Minimal Install as base

## Installation

dnf update -y
dnf install epel-release
dnf update
dnf module enable cobbler
dnf install cobbler
dnf install cobbler-web
dnf install pykickstart

dnf install mod_ssl
dnf install python3-librepo
dnf install syslinux
dnf install python3-ldap
dnf install rsync-daemon
dnf install dnsmasq

### SELINUX
setenforce 0
vi /etc/selinux/config and set to disabled

### DAEMONS
systemctl start rsyncd ; systemctl enable rsyncd
systemctl start cobblerd ; systemctl enable cobblerd
systemctl start httpd ; systemctl enable httpd
systemctl start tftp ; systemctl enable tftp
systemctl start dnsmasq ; systemctl enable dnsmasq

### FIREWALL
firewall-cmd --add-port=80/tcp --permanent
firewall-cmd --add-port=443/tcp --permanent
firewall-cmd --add-service=dhcp --permanent
firewall-cmd --add-port=69/tcp --permanent
firewall-cmd --add-port=69/udp --permanent
firewall-cmd --add-port=4011/udp --permanent
firewall-cmd --reload

### HTPPD
printf "foobar" | openssl dgst -sha3-512
then edit /etc/cobbler/users.digest

Not actually foobar of course

## COBBLER ITSELF

edit config file settings.yaml to use dnsmasq , dhcp, server line and next server line , tftpd and pxe once etc
remember it is .yaml now and requires TRUE not '1' , crypted password and such like

edit the modules.conf settings file
use _ not a . for dnsmasq and dhcp etc

edit the dnsmasq and dhcp settings files

We are a proxy and only provide the next server, boot file, once, to only known hosts

Question 1....
I did edit dhcp.template but as we aren't using it for dhcp I'm not certain that is necessary - I just edited it to be
the correct network

```
# Cobbler generated configuration file for dnsmasq
# This is how we are only a proxy - another dhcp server gives most of the information upon a mac addr broadcast for an IP etc
# $date

# Only for debugging
#log-dhcp

#  Be ONLY a proxy
dhcp-range=<our network>,proxy

# Only respond to clients that are known
dhcp-ignore=tag:!known

# Set this (and domain: see below) if you want to have a domain
# automatically added to simple names in a hosts-file.
expand-hosts
domain=<our domain>,<our network>

# What client architecture is it - legacy or EFI
# PXEClient:Arch:00007 - which means EFI (also 9 in some circumstances but not ours)
# 0 is legacy
# I tried dhcp set tag here for more readability which worked in 2.8.5 but not in 3.2.2 for some reason , digits do though

# Loads <tftp-root>/<something> from dnsmasq TFTP server depending on architecture to handle legacy and efi
pxe-service=0,"Boot Legacy (******      Cobbler controlled      ******)",grub/grub.0,$next_server
pxe-service=7,"Boot EFI (******      EFI Cobbler controlled      ******)",grub/grubx64.efi,$next_server

# Include the systems that must be "known"
$insert_cobbler_system_definitions
```

### RESTART
systemctl restart cobblerd
cobbler signature update
cobbler sync 

## DISTRO

dnf install wget
cd /home
wget http://mirror.bytemark.co.uk/centos/7.9.2009/isos/x86_64/CentOS-7-x86_64-DVD-2009.iso
mount -t iso9660 -o loop,ro /home/CentOS-7-x86_64-DVD-2009.iso /mnt
cobbler import --name=centos7 --arch=x86_64 --path=/mnt

Gets the right breed good...

```
task started: 2021-12-20_180954_import
task started (id=Media import, time=Mon Dec 20 18:09:54 2021)
running python triggers from /var/lib/cobbler/triggers/task/import/pre/*
running shell triggers from /var/lib/cobbler/triggers/task/import/pre/*
shell triggers finished successfully
Found a candidate signature: breed=redhat, version=rhel6
Found a candidate signature: breed=redhat, version=rhel7
Found a matching signature: breed=redhat, version=rhel7
Adding distros from path /var/www/cobbler/distro_mirror/centos7-x86_64:
creating new distro: centos7-x86_64
trying symlink: /var/www/cobbler/distro_mirror/centos7-x86_64 -> /var/www/cobbler/links/centos7-x86_64
creating new profile: centos7-x86_64
associating repos
checking for rsync repo(s)
checking for rhn repo(s)
checking for yum repo(s)
starting descent into /var/www/cobbler/distro_mirror/centos7-x86_64 for centos7-x86_64
processing repo at : /var/www/cobbler/distro_mirror/centos7-x86_64
need to process repo/comps: /var/www/cobbler/distro_mirror/centos7-x86_64
looking for /var/www/cobbler/distro_mirror/centos7-x86_64/repodata/*comps*.xml
Keeping repodata as-is :/var/www/cobbler/distro_mirror/centos7-x86_64/repodata
*** TASK COMPLETE ***
```

umount /mnt

## PROFILE and KICKSTART

Basically keyboard to UK and change the packages...to minimal or to graphical desktop

Change some naming of profile on web interface and make another one as a copy - Centos_Desktop , Centos_Minimal

Make 2 corresponding kickstart files

Associate with profiles on web interface

## SYSTEMS

Add Systems

cobbler system add \
  --name=testlegacy1 \
  --mac=my-mac-addr \
  --profile=Centos_Desktop \
  --interface=em1
  
cobbler system add \
--name=testlegacy2 \
--mac=my-mac-addr \
--profile=Centos_Minimal \
--interface=eth0 \

cobbler system add \
--name=test3 \
--mac=my-mac-addr \
--profile=Centos_Minimal \
--interface=eth0

cobbler sync

## MKGRUB and LOADERS

No loaders yet at all , cobbler get-loaders is deprecated according to what I have read

[root@cobbler loaders]# pwd
/var/lib/cobbler/loaders
[root@cobbler loaders]# ls
<nothing>

Run /usr/share/cobbler/bin/mkgrub.sh but export the right environment vars first

for these vars...
```
# DISTRO directory overrides. Pass these vars in from outside:
# export SYSLINUX_DIR=/usr/share/...;./mkgrub.sh
[[ -z "$SYSLINUX_DIR" ]] && SYSLINUX_DIR="/usr/share/syslinux"
[[ -z "$GRUB2_MOD_DIR" ]] && GRUB2_MOD_DIR="/usr/share/grub2"
```

SYSLINUX_DIR is ok

but /usr/share/grub2 does not exist - maybe something is not installed - No , turns out Centos puts it's grub2 stuff in /usr/lib/grub even though it is actually grub2..
```
# repoquery -l grub2-efi-x64-modules.noarch
/usr/lib/grub/x86_64-efi
/usr/lib/grub/x86_64-efi/acpi.mod
/usr/lib/grub/x86_64-efi/adler32.mod
...snip...
```

Lets run it and see - for some reason it DOES find the modules anyway (I didn't read all the script)

So various things are missing, for example for grub.0...for LEGACY
```
+ grub2-mkimage -O i386-pc-pxe -o /var/lib/cobbler/loaders/grub/grub.0 --prefix= all_video boot cat configfile echo true font gfxmenu gfxterm gzio halt iso9660 jpeg minicmd normal part_apple part_msdos part_gpt password_pbkdf2 png reboot search search_fs_uuid search_fs_file search_label sleep test video fat loadenv linux btrfs ext2 xfs jfs reiserfs tftp http luks gcry_rijndael gcry_sha1 gcry_sha256 mdraid09 mdraid1x lvm serial regexp tr chain pxe biosdisk
grub2-mkimage: error: cannot open `/usr/lib/grub/i386-pc/moddep.lst': No such file or directory.
```
and for grubx64 ...for EFI
```
+ grub2-mkimage -O x86_64-efi -o /var/lib/cobbler/loaders/grub/grubx64.efi --prefix= all_video boot cat configfile echo true font gfxmenu gfxterm gzio halt iso9660 jpeg minicmd normal part_apple part_msdos part_gpt password_pbkdf2 png reboot search search_fs_uuid search_fs_file search_label sleep test video fat loadenv linux btrfs ext2 xfs jfs reiserfs tftp http luks gcry_rijndael gcry_sha1 gcry_sha256 mdraid09 mdraid1x lvm serial regexp tr chain efinet
grub2-mkimage: error: cannot open `/usr/lib/grub/x86_64-efi/moddep.lst': No such file or directory.
```
This file is provided by the grub2-efi-x64-modules package...
```
[root@cobbler loaders]# dnf search grub2-efi-x64-modules
Last metadata expiration check: 3:16:39 ago on Mon 20 Dec 2021 16:04:46 GMT.
===================================================== Name Exactly Matched: grub2-efi-x64-modules ======================================================
grub2-efi-x64-modules.noarch : Modules used to build custom grub.efi images

[root@cobbler loaders]# dnf install grub2-efi-x64-modules
```
Lets try it now to see if it makes the EFI boot loader ok now then we can think about the legacy one
```
+ grub2-mkimage -O x86_64-efi -o /var/lib/cobbler/loaders/grub/grubx64.efi --prefix= all_video boot cat configfile echo true
font gfxmenu gfxterm gzio halt iso9660 jpeg minicmd normal part_apple part_msdos part_gpt password_pbkdf2 png reboot search
search_fs_uuid search_fs_file search_label sleep test video fat loadenv linux btrfs ext2 xfs jfs reiserfs tftp http luks gcry_rijndael
gcry_sha1 gcry_sha256 mdraid09 mdraid1x lvm serial regexp tr chain efinet
```
No issues now for this EFI.
Do the legacy basic one too
```
[root@cobbler loaders]# yum provides /usr/lib/grub/i386-pc/moddep.lst
Last metadata expiration check: 0:03:56 ago on Mon 20 Dec 2021 19:22:12 GMT.
grub2-pc-modules-1:2.02-106.el8.alma.noarch : Modules used to build custom grub images
Repo        : baseos
Matched from:
Filename    : /usr/lib/grub/i386-pc/moddep.lst
```
So lets install this package and try again for legacy
```
[root@cobbler loaders]# dnf install grub2-pc-modules
Last metadata expiration check: 0:05:05 ago on Mon 20 Dec 2021 19:22:12 GMT.
Dependencies resolved.
========================================================================================================================================================
 Package                                 Architecture                  Version                                      Repository                     Size
========================================================================================================================================================
Installing:
 grub2-pc-modules                        noarch                        1:2.02-106.el8.alma                          baseos                        915 k

Transaction Summary
========================================================================================================================================================
Install  1 Package

Total download size: 915 k
Installed size: 2.9 M
Is this ok [y/N]: y
Downloading Packages:
grub2-pc-modules-2.02-106.el8.alma.noarch.rpm                                                                           6.1 MB/s | 915 kB     00:00
--------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                   1.5 MB/s | 915 kB     00:00
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                1/1
  Installing       : grub2-pc-modules-1:2.02-106.el8.alma.noarch                                                                                    1/1
  Running scriptlet: grub2-pc-modules-1:2.02-106.el8.alma.noarch                                                                                    1/1
  Verifying        : grub2-pc-modules-1:2.02-106.el8.alma.noarch                                                                                    1/1

Installed:
  grub2-pc-modules-1:2.02-106.el8.alma.noarch

Complete!
```
What's mkgrub.sh like now for legacy?
```
+ grub2-mkimage -O i386-pc-pxe -o /var/lib/cobbler/loaders/grub/grub.0 --prefix= all_video boot cat configfile echo true font
 gfxmenu gfxterm gzio halt iso9660 jpeg minicmd normal part_apple part_msdos part_gpt password_pbkdf2 png reboot search search_fs_uuid
 search_fs_file search_label sleep test video fat loadenv linux btrfs ext2 xfs jfs reiserfs tftp http luks gcry_rijndael 
 gcry_sha1 gcry_sha256 mdraid09 mdraid1x lvm serial regexp tr chain pxe biosdisk
``` 
No issues with grub.0 now either

There are other architectures such as powerpc but we dont care about them

Now after a cobbler sync there are some loaders in the tftpboot dir
```
[root@cobbler loaders]# cd /var/lib/tftpboot/
[root@cobbler tftpboot]# ls
boot  etc  grub  grub.cfg  images  images2  ldlinux.c32  menu.c32  ppc  pxelinux.0  pxelinux.cfg  s390x
[root@cobbler tftpboot]# cd grub/
[root@cobbler grub]# ls
grub.0        grub.cfg      grubx64.efi  local_efi.cfg     local_powerpc-ieee1275.cfg  system_link
grubaa64.efi  grub.ppc64le  images       local_legacy.cfg  system                      x86_64_menu_items.cfg
[root@cobbler grub]# ls /var/lib/cobbler/loaders
grub  ldlinux.c32  menu.c32  pxelinux.0
[root@cobbler grub]# ls /var/lib/cobbler/loaders/grub
grub.0  grubaa64.efi  grub.ppc64le  grubx64.efi
```

Question 2...

Legacy now tries to load and run via grub.0 but bombs as it tries to read from fd0? (picture below in conversation)
Do I need to edit a grub template or something so it makes the correct grub.cfg file per host
Both legacy and EFI drop out to grub> prompt after attempting to boot.

Interestingly if I change to use pxelinux.0 for legacy directly it will boot and install fine but doesn't trigger the do it once
post install changes in cobbler and so goes into an install loop, this worked fine for legacy under 2.8.5

Question 3...

Where do I run mkgrub.sh?  -  on a centos7 machine or on the AlmaLinux 8.5 machine?

Question 4...
Having done this with some minor changes on both 2.8.5 (centos 7) and on 3.2.2 (AlmaLinux 8.5) , I'm fairly sure the grub that is being run although downloaded from the tftpboot server ok is not loading the associated grub.cfg.

What am I not doing?

**Comment by SchoolGuy** [comment]
Okay thank you for that very detailed post! My understanding (I am on the SUSE side of things) is that you need `mkgrub.sh` always because it creates the required GRUB bootloaders. Those loaders should be capable of both EFI and non-EFI booting. If they are not I would assume that either a GRUB module is missing or the configuration is wrong.

The legacy issue I have never seen and I don't think you have a floppy drive installed in your system, so I am out of ideas there.

Regarding the missing packages: Yes we are fully aware that for now this is counter intuitive and that is why `mkgrub.sh` was migrated from version 3.3.0 and newer to the `cobbler mkgrub`/`cobbler mkloaders` command which tries to help you in regard to this. Especially with 3.3.1 (not yet published) we did a number of improvements.

> Note: Don't unmount your ISO after a correct `cobbler import`! A `cobbler sync` is basically a "start-from-scratch" and that means that it will try to refetch the files from a now not existing location. I am aware of this issue and I try to fix it but for now those ISOs need to stay permanently mounted.

**Reply by Fluffyrodent** [reply]
Thank you for the response.  The leave ISO mounted is interesting so I will try that change first and report back.
Regards,
Stephen

**Comment by Fluffyrodent** [comment]
Hi,
So leaving the ISO mounted didn't alter the behaviour.  The mkgrub generated grub.0 and grubx64.efi are loaded as directed to in the dnsmasq template - I can tell this because the text line is echo'ed ok and the /var/log/messages shows it is being downloaded from tftpboot server , however it still then fails to load the appropriate grub cfg.

I'll keep reading but ideas would be great if anyone has seen this.

-Stephen.

**Reply by SchoolGuy** [reply]
So if `grub.0` and `grubx64.efi` are indeed downloaded to the client I am speechless.

However one straw I have left in my pocket: In theory you should see something before you drop into the EFI Shell. So what do you see? On the machines I had hand on in the past there was for a splitsecond an error message visible. This was in the past for me a missing GRUB module and/or other messages related to the reason why GRUB is aborting.

**Comment by Fluffyrodent** [comment]
Here's a snip of the grub.0 legacy boot error that flashes up , I haven't noticed one for the EFI grubx64.EFI one;
![grub 0 error on legacy cobbler](https://user-images.githubusercontent.com/69204611/149945983-43be9e0e-1adb-48d4-a885-17f48d013748.PNG)

**Reply by Fluffyrodent** [reply]
I think I might revert to 2.8.5 as then I can at least do legacy bare metal and hyperv installs okas I can just use pxelinux.0 and not grub.0.  Still need to work out how to get the EFI grub though for baremetal and hyperv grubx64.efi / shimx64 etc then though as there is no mkgrub.sh in that version

**Reply by SchoolGuy** [reply]
So all issue in regard to that issue I found is that your hard disk is failing. Maybe look at the SMART values (if the drive has them) and check if the drive is healthy?

**Reply by Fluffyrodent** [reply]
This is a HyperV VM and has no fd0 i.e. floppy drive.  I think I will drop this here as I can't spend much more time on it.  I have reverted the cobbler server to a centos7 machine and can boot legacy using plain pxelinux.0 from it ok.  As mentioned the EFI is another component on this architecture though.  If I can't figure it out for that I may switch to Foreman although I wanted to avoid that as it's overly complex for my needs.

Will advise if I get anywhere for your interest.

-Stephen.

**Reply by SchoolGuy** [reply]
I am deeply sorry that this happens. We have quite big testsuites which also to E2E testing. Personally all of the above works for me in the two environments I am managing with the help of Cobbler. Thus I am really out of ideas now... :/

**Reply by Fluffyrodent** [reply]
![grubx64_efi_error_centos7](https://user-images.githubusercontent.com/69204611/150149039-9eec155b-7ee9-47c5-9169-3a4482633b2b.PNG)

Above for reference.  This is an example of the grubx64.efi boot too after which it drops out to grub> prompt.

I did notice the centos7 method seems to have this BOOTX64.EFI involved somehow which I don't completely understand.  It is mentioned in Cobbler #1904 (that's an old post for Scientific Linux - but that's based on centos)

I shall keep reading.

**Comment by opoplawski** [comment]
I'm not clear about whether or not it tries to download a grub.cfg file and fails (which would be your issue) or that it doesn't try to download one at all.

**Reply by Fluffyrodent** [reply]
That's a good question.  After the grubx64 NBP is downloaded is when it bombs to "Grub>" prompt , and when I put the dhcp logging to ON in dnsmasq there is no additional information in /var/log/messages to indicate grub.cfg (or mac-addr file) download or not.

I did read somewhere in an old old post centos/redhat grub needed "grub.cfg-mac-addr" as the filename for config rather than straight mac-addr as a filename.  However I thought that was related to grub v1 and not grub2.  Also it was a really old post.

Tbh, I'm just really confused about grub now :(

**Reply by Fluffyrodent** [reply]
I wonder if grub can find the config file , I'm not sure it looks for something called after the mac address , rather than something called grub.cfg.  However the behaviour doesn't change if I copy the file name to grub.cfg or to the older grub.cfg-01-mac addr which I read about somewhere.

**Comment by Fluffyrodent** [comment]
I've updated the original post to make the situation on AlmaLinux 8.5 more clear and to highlight the questions I have better.

Does anyone have centos 7 auto installing succesfully with a grub.0 and grubx64.efi ?  If so can I see your dnsmasq and dhcp template and mkgrub process verbatim please? [not your IPs and names of course]

Regards,
Stephen.

**Comment by opoplawski** [comment]
I may have hit this same issue just now.  I was trying to setup a secure network boot configuration so I copied the almalinux shim.efi and grubx64.efi image from /boot/efi/EFI/almalinux.  After that change it would just boot the grub command prompt.  Running tshark I discovered that after loading grub/grub.cfg it started looking for other config file in EFI/almalinux because appartently the grub prefix variable was set to that.  I have so far just done a quick workaround of making that a link to the grub directory in the tftpboot directory.
